usr/lib/common/mech_list.c

```diff
@@ -44,39 +44,58 @@ CK_RV ock_generic_filter_mechanism_list(STDLL_TokData_t *tokdata,
 
 CK_RV ock_generic_get_mechanism_list(STDLL_TokData_t * tokdata,
                                      CK_MECHANISM_TYPE_PTR pMechanismList,
-                                     CK_ULONG_PTR pulCount)
+                                     CK_ULONG_PTR pulCount,
+                                     CK_BBOOL (*filter_mechanism)
+                                               (STDLL_TokData_t *tokdata,
+                                               CK_MECHANISM_TYPE mechanism))
 {
     int rc = CKR_OK;
-    unsigned int i;
-    if (pMechanismList == NULL) {
-        (*pulCount) = tokdata->mech_list_len;
-        goto out;
-    }
-    if ((*pulCount) < tokdata->mech_list_len) {
-        (*pulCount) = tokdata->mech_list_len;
-        TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
-        rc = CKR_BUFFER_TOO_SMALL;
-        goto out;
+    unsigned int i, j;
+
+    for (i = 0, j = 0; i < tokdata->mech_list_len; i++) {
+        if (filter_mechanism == NULL ||
+            filter_mechanism(tokdata, tokdata->mech_list[i].mech_type)) {
+            if (pMechanismList != NULL) {
+                if ((*pulCount) <= j) {
+                    (*pulCount) = j;
```

Comment:
Nice idea to combine anything into a single loop.
However, for the case where the buffer is too small to hold all mechs, the PKCS#11 standard states that it shold return in `*pulCount` the number of all mechanisms to be returned. 
The way you have it currently, you just return the current `j` value, but there might be more to return....

How about the following:

```
     for (i = 0, j = 0; i < tokdata->mech_list_len; i++) {
        if (filter_mechanism == NULL ||
            filter_mechanism(tokdata, tokdata->mech_list[i].mech_type)) {
            if (pMechanismList != NULL) {
                if ((*pulCount) <= j)
                    rc = CKR_BUFFER_TOO_SMALL;
                else
                    pMechanismList[j] = tokdata->mech_list[i].mech_type;
            }
            j++;
        }
    }

    (*pulCount) = j;
    
    if (rc == CKR_BUFFER_TOO_SMALL) {
        TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
    }
```

It keeps counting `j` even if the buffer is too small, but does not update the result list in that case.
I didn't want to print the trace message about buffer too small inside the loop, because it would then possibly print it multiple times, thus the check after the loop.